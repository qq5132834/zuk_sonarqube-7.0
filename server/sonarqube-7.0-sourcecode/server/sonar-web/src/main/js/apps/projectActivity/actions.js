/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
/*:: import type { Event } from './types'; */
/*:: import type { State } from './components/ProjectActivityAppContainer'; */

export const addCustomEvent = (analysis /*: string */, event /*: Event */) => (
  state /*: State */
) => ({
  analyses: state.analyses.map(item => {
    if (item.key !== analysis) {
      return item;
    }
    return { ...item, events: [...item.events, event] };
  })
});

export const deleteEvent = (analysis /*: string */, event /*: string */) => (
  state /*: State */
) => ({
  analyses: state.analyses.map(item => {
    if (item.key !== analysis) {
      return item;
    }
    return {
      ...item,
      events: item.events.filter(eventItem => eventItem.key !== event)
    };
  })
});

export const changeEvent = (analysis /*: string */, event /*: Event */) => (
  state /*: State */
) => ({
  analyses: state.analyses.map(item => {
    if (item.key !== analysis) {
      return item;
    }
    return {
      ...item,
      events: item.events.map(
        eventItem => (eventItem.key === event.key ? { ...eventItem, ...event } : eventItem)
      )
    };
  })
});

export const deleteAnalysis = (analysis /*: string */) => (state /*: State */) => ({
  analyses: state.analyses.filter(item => item.key !== analysis)
});
