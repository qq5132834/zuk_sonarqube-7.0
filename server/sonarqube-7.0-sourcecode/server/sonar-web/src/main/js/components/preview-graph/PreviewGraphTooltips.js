/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import BubblePopup from '../common/BubblePopup';
import DateFormatter from '../intl/DateFormatter';
import PreviewGraphTooltipsContent from './PreviewGraphTooltipsContent';
/*:: import type { Metric } from '../types'; */
/*:: import type { Serie } from '../charts/AdvancedTimeline'; */

/*::
type Props = {
  formatValue: (number | string) => string,
  graph: string,
  graphWidth: number,
  metrics: Array<Metric>,
  selectedDate: Date,
  series: Array<Serie & { translatedName: string }>,
  tooltipIdx: number,
  tooltipPos: number
};
*/

const TOOLTIP_WIDTH = 160;

export default class PreviewGraphTooltips extends React.PureComponent {
  /*:: props: Props; */

  render() {
    const { tooltipIdx } = this.props;
    const top = 16;
    let left = this.props.tooltipPos;
    let customClass;
    if (left > this.props.graphWidth - TOOLTIP_WIDTH) {
      left -= TOOLTIP_WIDTH;
      customClass = 'bubble-popup-right';
    }

    return (
      <BubblePopup customClass={customClass} position={{ top, left, width: TOOLTIP_WIDTH }}>
        <div className="overview-analysis-graph-tooltip">
          <div className="overview-analysis-graph-tooltip-title">
            <DateFormatter date={this.props.selectedDate} long={true} />
          </div>
          <table className="width-100">
            <tbody>
              {this.props.series.map((serie, idx) => {
                const point = serie.data[tooltipIdx];
                if (!point || (!point.y && point.y !== 0)) {
                  return null;
                }
                return (
                  <PreviewGraphTooltipsContent
                    key={serie.name}
                    style={idx.toString()}
                    translatedName={serie.translatedName}
                    value={this.props.formatValue(point.y)}
                  />
                );
              })}
            </tbody>
          </table>
        </div>
      </BubblePopup>
    );
  }
}
