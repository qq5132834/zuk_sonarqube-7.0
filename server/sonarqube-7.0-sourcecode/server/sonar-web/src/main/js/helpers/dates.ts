/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  differenceInDays as _differenceInDays,
  differenceInSeconds as _differenceInSeconds,
  differenceInYears as _differenceInYears,
  isSameDay as _isSameDay,
  parse,
  startOfDay as _startOfDay
} from 'date-fns';

function pad(number: number) {
  if (number < 10) {
    return '0' + number;
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function startOfDay(date: Date): Date {
  return _startOfDay(date);
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
  return _isSameDay(dateLeft, dateRight);
}

export function differenceInYears(dateLeft: Date, dateRight: Date): number {
  return _differenceInYears(dateLeft, dateRight);
}

export function differenceInDays(dateLeft: Date, dateRight: Date): number {
  return _differenceInDays(dateLeft, dateRight);
}

export function differenceInSeconds(dateLeft: Date, dateRight: Date): number {
  return _differenceInSeconds(dateLeft, dateRight);
}
