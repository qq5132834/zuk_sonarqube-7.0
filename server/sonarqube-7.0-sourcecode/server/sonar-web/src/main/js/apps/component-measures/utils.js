/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { groupBy, memoize, sortBy, toPairs } from 'lodash';
import { getLocalizedMetricName } from '../../helpers/l10n';
import { cleanQuery, parseAsString, serializeString } from '../../helpers/query';
import { domains } from './config/domains';
import { bubbles } from './config/bubbles';
import { enhanceMeasure } from '../../components/measure/utils';
/*:: import type { Component, ComponentEnhanced, Query } from './types'; */
/*:: import type { RawQuery } from '../../helpers/query'; */
/*:: import type { Metric } from '../../store/metrics/actions'; */
/*:: import type { MeasureEnhanced } from '../../components/measure/types'; */

export const PROJECT_OVERVEW = 'project_overview';
export const DEFAULT_VIEW = 'list';
export const DEFAULT_METRIC = PROJECT_OVERVEW;
export const KNOWN_DOMAINS = [
  'Releasability',
  'Reliability',
  'Security',
  'Maintainability',
  'Coverage',
  'Duplications',
  'Size',
  'Complexity'
];
const BANNED_MEASURES = [
  'blocker_violations',
  'new_blocker_violations',
  'critical_violations',
  'new_critical_violations',
  'major_violations',
  'new_major_violations',
  'minor_violations',
  'new_minor_violations',
  'info_violations',
  'new_info_violations'
];

export function filterMeasures(
  measures /*: Array<MeasureEnhanced> */
) /*: Array<MeasureEnhanced> */ {
  return measures.filter(measure => !BANNED_MEASURES.includes(measure.metric.key));
}

export function sortMeasures(
  domainName /*: string */,
  measures /*: Array<MeasureEnhanced | string> */
) /*: Array<MeasureEnhanced | string> */ {
  const config = domains[domainName] || {};
  const configOrder = config.order || [];
  return sortBy(measures, [
    item => {
      if (typeof item === 'string') {
        return configOrder.indexOf(item);
      }
      const idx = configOrder.indexOf(item.metric.key);
      return idx >= 0 ? idx : configOrder.length;
    },
    item => (typeof item === 'string' ? item : getLocalizedMetricName(item.metric))
  ]);
}

export function addMeasureCategories(
  domainName /*: string */,
  measures /*: Array<MeasureEnhanced> */
) /*: Array<any> */ {
  const categories = domains[domainName] && domains[domainName].categories;
  if (categories && categories.length > 0) {
    return [...categories, ...measures];
  }
  return measures;
}

export function enhanceComponent(
  component /*: Component */,
  metric /*: ?Metric */,
  metrics /*: { [string]: Metric } */
) /*: ComponentEnhanced */ {
  const enhancedMeasures = component.measures.map(measure => enhanceMeasure(measure, metrics));
  // $FlowFixMe metric can't be null since there is a guard for it
  const measure = metric && enhancedMeasures.find(measure => measure.metric.key === metric.key);
  const value = measure ? measure.value : null;
  const leak = measure ? measure.leak : null;
  return { ...component, value, leak, measures: enhancedMeasures };
}

export function isFileType(component /*: Component */) /*: boolean */ {
  return ['FIL', 'UTS'].includes(component.qualifier);
}

export function isViewType(component /*: Component */) /*: boolean */ {
  return ['VW', 'SVW', 'APP'].includes(component.qualifier);
}

export const groupByDomains = memoize((measures /*: Array<MeasureEnhanced> */) => {
  const domains = toPairs(groupBy(measures, measure => measure.metric.domain)).map(r => ({
    name: r[0],
    measures: r[1]
  }));

  return sortBy(domains, [
    domain => {
      const idx = KNOWN_DOMAINS.indexOf(domain.name);
      return idx >= 0 ? idx : KNOWN_DOMAINS.length;
    },
    'name'
  ]);
});

export function getDefaultView(metric /*: string */) /*: string */ {
  if (!hasList(metric)) {
    return 'tree';
  }
  return DEFAULT_VIEW;
}

export function hasList(metric /*: string */) /*: boolean */ {
  return !['releasability_rating', 'releasability_effort'].includes(metric);
}

export function hasTree(metric /*: string */) /*: boolean */ {
  return metric !== 'alert_status';
}

export function hasTreemap(metric /*: string */, type /*: string */) /*: boolean */ {
  return ['PERCENT', 'RATING', 'LEVEL'].includes(type) && hasTree(metric);
}

export function hasBubbleChart(domainName /*: string */) /*: boolean */ {
  return bubbles[domainName] != null;
}

export function hasFacetStat(metric /*: string */) /*: boolean */ {
  return metric !== 'alert_status';
}

export function getBubbleMetrics(domain /*: string */, metrics /*: { [string]: Metric } */) {
  const conf = bubbles[domain];
  return {
    x: metrics[conf.x],
    y: metrics[conf.y],
    size: metrics[conf.size],
    colors: conf.colors ? conf.colors.map(color => metrics[color]) : null
  };
}

export function getBubbleYDomain(domain /*: string */) {
  return bubbles[domain].yDomain;
}

export function isProjectOverview(metric /*: string */) {
  return metric === PROJECT_OVERVEW;
}

const parseView = memoize((rawView /*:: ? */ /*: string */, metric /*: string */) => {
  const view = parseAsString(rawView) || DEFAULT_VIEW;
  if (!hasTree(metric)) {
    return 'list';
  } else if (view === 'list' && !hasList(metric)) {
    return 'tree';
  }
  return view;
});

export const parseQuery = memoize((urlQuery /*: RawQuery */) => {
  const metric = parseAsString(urlQuery['metric']) || DEFAULT_METRIC;
  return {
    metric,
    selected: parseAsString(urlQuery['selected']),
    view: parseView(urlQuery['view'], metric)
  };
});

export const serializeQuery = memoize((query /*: Query */) => {
  return cleanQuery({
    metric: query.metric === DEFAULT_METRIC ? null : serializeString(query.metric),
    selected: serializeString(query.selected),
    view: query.view === DEFAULT_VIEW ? null : serializeString(query.view)
  });
});
