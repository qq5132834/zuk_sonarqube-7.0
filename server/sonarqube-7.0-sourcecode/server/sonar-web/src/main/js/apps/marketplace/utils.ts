/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { memoize, sortBy } from 'lodash';
import { Plugin, PluginAvailable, PluginInstalled, PluginPending } from '../../api/plugins';
import { cleanQuery, parseAsString, RawQuery, serializeString } from '../../helpers/query';
import { Edition } from '../../api/marketplace';

export interface Query {
  filter: string;
  search?: string;
}

export function isPluginAvailable(plugin: Plugin): plugin is PluginAvailable {
  return (plugin as any).release !== undefined;
}

export function isPluginInstalled(plugin: Plugin): plugin is PluginInstalled {
  return isPluginPending(plugin) && (plugin as any).updatedAt !== undefined;
}

export function isPluginPending(plugin: Plugin): plugin is PluginPending {
  return (plugin as any).version !== undefined;
}

export function filterPlugins(plugins: Plugin[], search: string): Plugin[] {
  const s = search.toLowerCase();
  return plugins.filter(plugin => {
    return (
      plugin.name.toLowerCase().includes(s) ||
      plugin.description.toLowerCase().includes(s) ||
      (plugin.category || '').toLowerCase().includes(s)
    );
  });
}

const EDITIONS_ORDER = ['community', 'developer', 'enterprise', 'datacenter'];
export function sortEditions(editions: Edition[]): Edition[] {
  return sortBy(editions, edition => EDITIONS_ORDER.indexOf(edition.key));
}

export const DEFAULT_FILTER = 'all';
export const parseQuery = memoize((urlQuery: RawQuery): Query => ({
  filter: parseAsString(urlQuery['filter']) || DEFAULT_FILTER,
  search: parseAsString(urlQuery['search'])
}));

export const serializeQuery = memoize((query: Query): RawQuery =>
  cleanQuery({
    filter: query.filter === DEFAULT_FILTER ? undefined : serializeString(query.filter),
    search: query.search ? serializeString(query.search) : undefined
  })
);
