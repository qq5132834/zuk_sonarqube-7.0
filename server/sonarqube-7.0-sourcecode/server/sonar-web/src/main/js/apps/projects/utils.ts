/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { sumBy, uniq } from 'lodash';
import { translate } from '../../helpers/l10n';
import { RequestData } from '../../helpers/request';
import { getOrganizations } from '../../api/organizations';
import { searchProjects, Facet } from '../../api/components';
import { getMeasuresForProjects } from '../../api/measures';
import { isDiffMetric, getPeriodValue } from '../../helpers/measures';
import { Query, convertToFilter } from './query';

interface SortingOption {
  class?: string;
  value: string;
}

export const SORTING_METRICS: SortingOption[] = [
  { value: 'name' },
  { value: 'analysis_date' },
  { value: 'reliability' },
  { value: 'security' },
  { value: 'maintainability' },
  { value: 'coverage' },
  { value: 'duplications' },
  { value: 'size' }
];

export const SORTING_LEAK_METRICS: SortingOption[] = [
  { value: 'name' },
  { value: 'analysis_date' },
  { value: 'new_reliability', class: 'projects-leak-sorting-option' },
  { value: 'new_security', class: 'projects-leak-sorting-option' },
  { value: 'new_maintainability', class: 'projects-leak-sorting-option' },
  { value: 'new_coverage', class: 'projects-leak-sorting-option' },
  { value: 'new_duplications', class: 'projects-leak-sorting-option' },
  { value: 'new_lines', class: 'projects-leak-sorting-option' }
];

export const SORTING_SWITCH: { [x: string]: string } = {
  analysis_date: 'analysis_date',
  name: 'name',
  reliability: 'new_reliability',
  security: 'new_security',
  maintainability: 'new_maintainability',
  coverage: 'new_coverage',
  duplications: 'new_duplications',
  size: 'new_lines',
  new_reliability: 'reliability',
  new_security: 'security',
  new_maintainability: 'maintainability',
  new_coverage: 'coverage',
  new_duplications: 'duplications',
  new_lines: 'size'
};

export const VIEWS = ['overall', 'leak'];

export const VISUALIZATIONS = [
  'risk',
  'reliability',
  'security',
  'maintainability',
  'coverage',
  'duplications'
];

const PAGE_SIZE = 50;
const PAGE_SIZE_VISUALIZATIONS = 99;

const METRICS = [
  'alert_status',
  'bugs',
  'reliability_rating',
  'vulnerabilities',
  'security_rating',
  'code_smells',
  'sqale_rating',
  'duplicated_lines_density',
  'coverage',
  'ncloc',
  'ncloc_language_distribution'
];

const LEAK_METRICS = [
  'alert_status',
  'new_bugs',
  'new_reliability_rating',
  'new_vulnerabilities',
  'new_security_rating',
  'new_code_smells',
  'new_maintainability_rating',
  'new_coverage',
  'new_duplicated_lines_density',
  'new_lines'
];

const METRICS_BY_VISUALIZATION: { [x: string]: string[] } = {
  risk: ['reliability_rating', 'security_rating', 'coverage', 'ncloc', 'sqale_index'],
  // x, y, size, color
  reliability: ['ncloc', 'reliability_remediation_effort', 'bugs', 'reliability_rating'],
  security: ['ncloc', 'security_remediation_effort', 'vulnerabilities', 'security_rating'],
  maintainability: ['ncloc', 'sqale_index', 'code_smells', 'sqale_rating'],
  coverage: ['complexity', 'coverage', 'uncovered_lines'],
  duplications: ['ncloc', 'duplicated_lines_density', 'duplicated_blocks']
};

const FACETS = [
  'reliability_rating',
  'security_rating',
  'sqale_rating',
  'coverage',
  'duplicated_lines_density',
  'ncloc',
  'alert_status',
  'languages',
  'tags'
];

const LEAK_FACETS = [
  'new_reliability_rating',
  'new_security_rating',
  'new_maintainability_rating',
  'new_coverage',
  'new_duplicated_lines_density',
  'new_lines',
  'alert_status',
  'languages',
  'tags'
];

const CUMULATIVE_FACETS = [
  'reliability',
  'new_reliability',
  'security',
  'new_security',
  'maintainability',
  'new_maintainability',
  'coverage',
  'new_coverage',
  'duplications',
  'new_duplications',
  'size',
  'new_lines'
];

const REVERSED_FACETS = ['coverage', 'new_coverage'];

export function localizeSorting(sort?: string): string {
  return translate('projects.sort', sort || 'name');
}

export function parseSorting(sort: string): { sortValue: string; sortDesc: boolean } {
  const desc = sort[0] === '-';
  return { sortValue: desc ? sort.substr(1) : sort, sortDesc: desc };
}

export function fetchProjects(
  query: Query,
  isFavorite: boolean,
  organization?: string,
  pageIndex = 1
) {
  const ps = query.view === 'visualizations' ? PAGE_SIZE_VISUALIZATIONS : PAGE_SIZE;
  const data = convertToQueryData(query, isFavorite, organization, {
    p: pageIndex > 1 ? pageIndex : undefined,
    ps,
    facets: defineFacets(query).join(),
    f: 'analysisDate,leakPeriodDate'
  });
  return searchProjects(data).then(({ components, facets, paging }) => {
    return Promise.all([
      fetchProjectMeasures(components, query),
      fetchProjectOrganizations(components)
    ]).then(([measures, organizations]) => {
      return {
        facets: getFacetsMap(facets),
        projects: components
          .map(component => {
            const componentMeasures: { [key: string]: string } = {};
            measures.filter(measure => measure.component === component.key).forEach(measure => {
              const value = isDiffMetric(measure.metric)
                ? getPeriodValue(measure, 1)
                : measure.value;
              if (value !== undefined) {
                componentMeasures[measure.metric] = value;
              }
            });
            return { ...component, measures: componentMeasures };
          })
          .map(component => {
            const organization = organizations.find(o => o.key === component.organization);
            return { ...component, organization };
          }),
        total: paging.total
      };
    });
  });
}

function defineMetrics(query: Query): string[] {
  switch (query.view) {
    case 'visualizations':
      return METRICS_BY_VISUALIZATION[query.visualization || 'risk'];
    case 'leak':
      return LEAK_METRICS;
    default:
      return METRICS;
  }
}

function defineFacets(query: Query): string[] {
  if (query.view === 'leak') {
    return LEAK_FACETS;
  }
  return FACETS;
}

function convertToQueryData(
  query: Query,
  isFavorite: boolean,
  organization?: string,
  defaultData = {}
) {
  const data: RequestData = { ...defaultData, organization };
  const filter = convertToFilter(query, isFavorite);
  const sort = convertToSorting(query as any);

  if (filter) {
    data.filter = filter;
  }
  if (sort.s) {
    data.s = sort.s;
  }
  if (sort.asc !== undefined) {
    data.asc = sort.asc;
  }
  return data;
}

function fetchProjectMeasures(projects: Array<{ key: string }>, query: Query) {
  if (!projects.length) {
    return Promise.resolve([]);
  }

  const projectKeys = projects.map(project => project.key);
  const metrics = defineMetrics(query);
  return getMeasuresForProjects(projectKeys, metrics);
}

function fetchProjectOrganizations(projects: Array<{ organization: string }>) {
  if (!projects.length) {
    return Promise.resolve([]);
  }

  const organizations = uniq(projects.map(project => project.organization));
  return getOrganizations({ organizations: organizations.join() }).then(r => r.organizations);
}

function mapFacetValues(values: Array<{ val: string; count: number }>) {
  const map: { [value: string]: number } = {};
  values.forEach(value => {
    map[value.val] = value.count;
  });
  return map;
}

export function cumulativeMapFacetValues(values: Array<{ val: string; count: number }>) {
  const noDataVal = values.find(value => value.val === 'NO_DATA');
  const filteredValues = noDataVal ? values.filter(value => value.val !== 'NO_DATA') : values;

  let sum = sumBy(filteredValues, value => value.count);
  const map: { [value: string]: number } = {};
  filteredValues.forEach((value, index) => {
    map[value.val] = index > 0 && index < values.length - 1 ? sum : value.count;
    sum -= value.count;
  });

  if (noDataVal) {
    map[noDataVal.val] = noDataVal.count;
  }
  return map;
}

function getFacetsMap(facets: Facet[]) {
  const map: { [property: string]: { [value: string]: number } } = {};
  facets.forEach(facet => {
    const property = mapMetricToProperty(facet.property);
    const { values } = facet;
    if (REVERSED_FACETS.includes(property)) {
      values.reverse();
    }
    map[property] = CUMULATIVE_FACETS.includes(property)
      ? cumulativeMapFacetValues(values)
      : mapFacetValues(values);
  });
  return map;
}

function mapPropertyToMetric(property?: string) {
  const map: { [property: string]: string } = {
    analysis_date: 'analysisDate',
    reliability: 'reliability_rating',
    new_reliability: 'new_reliability_rating',
    security: 'security_rating',
    new_security: 'new_security_rating',
    maintainability: 'sqale_rating',
    new_maintainability: 'new_maintainability_rating',
    coverage: 'coverage',
    new_coverage: 'new_coverage',
    duplications: 'duplicated_lines_density',
    new_duplications: 'new_duplicated_lines_density',
    size: 'ncloc',
    new_lines: 'new_lines',
    gate: 'alert_status',
    languages: 'languages',
    tags: 'tags',
    search: 'query'
  };
  return property && map[property];
}

function convertToSorting({ sort }: Query): { s?: string; asc?: boolean } {
  if (sort && sort[0] === '-') {
    return { s: mapPropertyToMetric(sort.substr(1)), asc: false };
  }
  return { s: mapPropertyToMetric(sort) };
}

function mapMetricToProperty(metricKey: string) {
  const map: { [metric: string]: string } = {
    analysisDate: 'analysis_date',
    reliability_rating: 'reliability',
    new_reliability_rating: 'new_reliability',
    security_rating: 'security',
    new_security_rating: 'new_security',
    sqale_rating: 'maintainability',
    new_maintainability_rating: 'new_maintainability',
    coverage: 'coverage',
    new_coverage: 'new_coverage',
    duplicated_lines_density: 'duplications',
    new_duplicated_lines_density: 'new_duplications',
    ncloc: 'size',
    new_lines: 'new_lines',
    alert_status: 'gate',
    languages: 'languages',
    tags: 'tags',
    query: 'search'
  };
  return map[metricKey];
}
