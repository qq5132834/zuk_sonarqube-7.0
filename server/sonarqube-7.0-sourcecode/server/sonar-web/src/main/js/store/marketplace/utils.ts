/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { sortBy } from 'lodash';
import { Edition, EditionsPerVersion } from '../../api/marketplace';

export function getEditionsForLastVersion(editions: EditionsPerVersion): Edition[] {
  const sortedVersion = sortBy(Object.keys(editions), [
    (version: string) => -Number(version.split('.')[0]),
    (version: string) => -Number(version.split('.')[1] || 0),
    (version: string) => -Number(version.split('.')[2] || 0)
  ]);
  return editions[sortedVersion[0]];
}

export function getEditionsForVersion(
  editions: EditionsPerVersion,
  version: string
): Edition[] | undefined {
  const minorVersion = version.match(/\d+\.\d+.\d+/);
  if (minorVersion) {
    if (editions[minorVersion[0]]) {
      return editions[minorVersion[0]];
    }
  }
  const majorVersion = version.match(/\d+\.\d+/);
  if (majorVersion) {
    if (editions[majorVersion[0]]) {
      return editions[majorVersion[0]];
    }
  }
  return undefined;
}
