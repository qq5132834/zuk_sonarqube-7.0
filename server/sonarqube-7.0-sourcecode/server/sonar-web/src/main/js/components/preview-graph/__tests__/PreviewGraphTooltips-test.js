/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import PreviewGraphTooltips from '../PreviewGraphTooltips';
import { DEFAULT_GRAPH } from '../../../apps/projectActivity/utils';
import { parseDate } from '../../../helpers/dates';

const SERIES_ISSUES = [
  {
    name: 'code_smells',
    data: [
      {
        x: '2011-10-01T22:01:00.000Z',
        y: 18
      },
      {
        x: '2011-10-25T10:27:41.000Z',
        y: 15
      }
    ],
    translatedName: 'Code Smells'
  },
  {
    name: 'bugs',
    data: [
      {
        x: '2011-10-01T22:01:00.000Z',
        y: 3
      },
      {
        x: '2011-10-25T10:27:41.000Z',
        y: 0
      }
    ],
    translatedName: 'Bugs'
  },
  {
    name: 'vulnerabilities',
    data: [
      {
        x: '2011-10-01T22:01:00.000Z',
        y: 0
      },
      {
        x: '2011-10-25T10:27:41.000Z',
        y: 1
      }
    ],
    translatedName: 'Vulnerabilities'
  }
];

const METRICS = [
  { key: 'code_smells', name: 'Code Smells', type: 'INT' },
  { key: 'bugs', name: 'Bugs', type: 'INT' },
  { key: 'vulnerabilities', name: 'Vulnerabilities', type: 'INT', custom: true }
];

const DEFAULT_PROPS = {
  formatValue: val => 'Formated.' + val,
  graph: DEFAULT_GRAPH,
  graphWidth: 150,
  metrics: METRICS,
  selectedDate: parseDate('2011-10-01T22:01:00.000Z'),
  series: SERIES_ISSUES,
  tooltipIdx: 0,
  tooltipPos: 25
};

it('should render correctly', () => {
  expect(
    shallow(
      <PreviewGraphTooltips
        {...DEFAULT_PROPS}
        graph="random"
        selectedDate={parseDate('2011-10-25T10:27:41.000Z')}
        tooltipIdx={1}
      />
    )
  ).toMatchSnapshot();
});
