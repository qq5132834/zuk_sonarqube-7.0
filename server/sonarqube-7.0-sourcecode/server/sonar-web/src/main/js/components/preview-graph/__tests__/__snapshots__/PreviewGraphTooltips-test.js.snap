// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<BubblePopup
  customClass="bubble-popup-right"
  position={
    Object {
      "left": -135,
      "top": 16,
      "width": 160,
    }
  }
>
  <div
    className="overview-analysis-graph-tooltip"
  >
    <div
      className="overview-analysis-graph-tooltip-title"
    >
      <DateFormatter
        date={2011-10-25T10:27:41.000Z}
        long={true}
      />
    </div>
    <table
      className="width-100"
    >
      <tbody>
        <PreviewGraphTooltipsContent
          key="code_smells"
          style="0"
          translatedName="Code Smells"
          value="Formated.15"
        />
        <PreviewGraphTooltipsContent
          key="bugs"
          style="1"
          translatedName="Bugs"
          value="Formated.0"
        />
        <PreviewGraphTooltipsContent
          key="vulnerabilities"
          style="2"
          translatedName="Vulnerabilities"
          value="Formated.1"
        />
      </tbody>
    </table>
  </div>
</BubblePopup>
`;
